boa/src/value/to_integer_or_infinity.rs

```diff
@@ -0,0 +1,99 @@
+use super::*;
+
+use std::convert::TryFrom;
+
+/// Represents the result of ToIntegerOrInfinity operation
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum IntegerOrInfinity {
+    Integer(isize),
+    Undefined,
+    PositiveInfinity,
+    NegativeInfinity,
+}
+
+impl IntegerOrInfinity {
+    /// Represents the algorithm to calculate `relativeStart` (or `k`) in array functions.
+    pub fn as_relative_start(self, len: usize) -> usize {
+        match self {
+            // 1. If relativeStart is -∞, let k be 0.
+            IntegerOrInfinity::NegativeInfinity => 0,
+            // 2. Else if relativeStart < 0, let k be max(len + relativeStart, 0).
+            IntegerOrInfinity::Integer(i) if i < 0 => Self::offset(len, i),
+            // 3. Else, let k be min(relativeStart, len).
+            IntegerOrInfinity::Integer(i) => (i as usize).min(len),
+
+            // Special case - postive infinity. `len` is always smaller than +inf, thus from (3)
+            IntegerOrInfinity::PositiveInfinity => len,
+            // Special case - `undefined` is treated like 0
+            IntegerOrInfinity::Undefined => 0,
+        }
+    }
+
+    /// Represents the algorithm to calculate `relativeEnd` (or `final`) in array functions.
+    pub fn as_relative_end(self, len: usize) -> usize {
+        match self {
+            // 1. If end is undefined, let relativeEnd be len
+            IntegerOrInfinity::Undefined => len,
+
+            // 1. cont, else let relativeEnd be ? ToIntegerOrInfinity(end).
+
+            // 2. If relativeEnd is -∞, let final be 0.
+            IntegerOrInfinity::NegativeInfinity => 0,
+            // 3. Else if relativeEnd < 0, let final be max(len + relativeEnd, 0).
+            IntegerOrInfinity::Integer(i) if i < 0 => Self::offset(len, i),
+            // 4. Else, let final be min(relativeEnd, len).
+            IntegerOrInfinity::Integer(i) => (i as usize).min(len),
+
+            // Special case - postive infinity. `len` is always smaller than +inf, thus from (4)
+            IntegerOrInfinity::PositiveInfinity => len,
+        }
+    }
+
+    fn offset(len: usize, i: isize) -> usize {
+        debug_assert!(i < 0);
+        if i == isize::MIN {
+            len.saturating_sub(isize::MAX as usize).saturating_sub(1)
+        } else {
+            len.saturating_sub(i.saturating_neg() as usize)
+        }
+    }
+}
+
+impl Value {
+    /// Converts argument to an integer, +∞, or -∞.
+    ///
+    /// See: <https://tc39.es/ecma262/#sec-tointegerorinfinity>
+    pub fn to_integer_or_infinity(&self, context: &mut Context) -> Result<IntegerOrInfinity> {
+        // Special case - `undefined`
+        if self.is_undefined() {
+            return Ok(IntegerOrInfinity::Undefined);
+        }
```

Comment:
Why do we treat undefined as a special case in the spec https://tc39.es/ecma262/#sec-tointegerorinfinity the first step is `ToNumber()` call.